# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather.csv` file. 

#### First import it into a data frame called `austin`.

In [2]:
# Your code here
austin=pd.read_csv('../austin_weather.csv')

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
# Your code here
display(austin.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1319 non-null object
DewPointAvgF                  1319 non-null object
DewPointLowF                  1319 non-null object
HumidityHighPercent           1319 non-null object
HumidityAvgPercent            1319 non-null object
HumidityLowPercent            1319 non-null object
SeaLevelPressureHighInches    1319 non-null object
SeaLevelPressureAvgInches     1319 non-null object
SeaLevelPressureLowInches     1319 non-null object
VisibilityHighMiles           1319 non-null object
VisibilityAvgMiles            1319 non-null object
VisibilityLowMiles            1319 non-null object
WindHighMPH                   1319 non-null object
WindAvgMPH                   

None

In [4]:
# Your code here
display(austin.describe(include='all'))

Date    TempHighF     TempAvgF     TempLowF DewPointHighF  \
count         1319  1319.000000  1319.000000  1319.000000          1319   
unique        1319          NaN          NaN          NaN            64   
top     2014-11-21          NaN          NaN          NaN            74   
freq             1          NaN          NaN          NaN           112   
mean           NaN    80.862775    70.642911    59.902957           NaN   
std            NaN    14.766523    14.045904    14.190648           NaN   
min            NaN    32.000000    29.000000    19.000000           NaN   
25%            NaN    72.000000    62.000000    49.000000           NaN   
50%            NaN    83.000000    73.000000    63.000000           NaN   
75%            NaN    92.000000    83.000000    73.000000           NaN   
max            NaN   107.000000    93.000000    81.000000           NaN   

       DewPointAvgF DewPointLowF HumidityHighPercent HumidityAvgPercent  \
count          1319         1319                1319               1319   
unique           66           73                  58                 69   
top              71           66                 100                 64   
freq             84           62                 221                 51   
mean            NaN          NaN                 NaN                NaN   
std             NaN          NaN                 NaN                NaN   
min             NaN          NaN                 NaN                NaN   
25%             NaN          NaN                 NaN                NaN   
50%             NaN          NaN                 NaN                NaN   
75%             NaN          NaN                 NaN                NaN   
max             NaN          NaN                 NaN                NaN   

       HumidityLowPercent  ... SeaLevelPressureAvgInches  \
count                1319  ...                      1319   
unique                 82  ...                       101   
top                    52  ...                     29.95   
freq                   46  ...                        49   
mean                  NaN  ...                       NaN   
std                   NaN  ...                       NaN   
min                   NaN  ...                       NaN   
25%                   NaN  ...                       NaN   
50%                   NaN  ...                       NaN   
75%                   NaN  ...                       NaN   
max                   NaN  ...                       NaN   

       SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
count                       1319                1319               1319   
unique                       105                   5                 10   
top                        29.88                  10                 10   
freq                          48                1302                859   
mean                         NaN                 NaN                NaN   
std                          NaN                 NaN                NaN   
min                          NaN                 NaN                NaN   
25%                          NaN                 NaN                NaN   
50%                          NaN                 NaN                NaN   
75%                          NaN                 NaN                NaN   
max                          NaN                 NaN                NaN   

       VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
count                1319        1319       1319        1319   
unique                 12          22         13          37   
top                    10          13          4          20   
freq                  611         212        239         122   
mean                  NaN         NaN        NaN         NaN   
std                   NaN         NaN        NaN         NaN   
min                   NaN         NaN        NaN         NaN   
25%                   NaN         NaN        NaN         NaN   
50%           

In [5]:
# Your code here
pd.options.display.max_columns = None
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

1)The majority of the columns that should be of int64 type, are object type; these should be changed. 
2)The events column seems to have little data, and its information is string type
3)Date is not in time-date mode. 

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [6]:
# Your code here
austin['DewPointAvgF'].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

# Your observation here
There is a character, '-', that it is not a number. This can be an error or a null value; we should take this out so the rest can be treated as numbers. 

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [7]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [8]:
# Your code here
austin[wrong_type_columns]=austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [9]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1312 non-null float64
DewPointAvgF                  1312 non-null float64
DewPointLowF                  1312 non-null float64
HumidityHighPercent           1317 non-null float64
HumidityAvgPercent            1317 non-null float64
HumidityLowPercent            1317 non-null float64
SeaLevelPressureHighInches    1316 non-null float64
SeaLevelPressureAvgInches     1316 non-null float64
SeaLevelPressureLowInches     1316 non-null float64
VisibilityHighMiles           1307 non-null float64
VisibilityAvgMiles            1307 non-null float64
VisibilityLowMiles            1307 non-null float64
WindHighMPH                   1317 non-null float64
WindAvgMPH      

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [10]:
# Your code here
#for index,t in 
l1=austin.isnull().any(axis=1).index.tolist()
l2=austin.isnull().any(axis=1).tolist()

m=[a for a, b in zip(l1, l2) if b == True]


missing_values=austin.loc[m,:]

missing_values

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4     2013-12-25         58        50        41           44.0          40.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
42    2014-02-01         76        66        55           62.0          59.0   
51    2014-02-10         60        48        35           49.0          36.0   
66    2014-02-25         71        62        52           65.0          60.0   
95    2014-03-26         66        60        54           55.0          47.0   
102   2014-04-02         82        76        70           69.0          67.0   
103   2014-04-03         82        77        71           69.0          67.0   
104   2014-04-04         74        64        54           69.0          35.0   
105   2014-04-05         65        60        54           46.0          38.0   
113   2014-04-13         78        73        68           68.0          66.0   
126   2014-04-26         86        77        68           69.0          66.0   
139   2014-05-09         86        76        66           71.0          68.0   
152   2014-05-22         87        79        70           68.0          65.0   
154   2014-05-24         86        78        70           70.0          68.0   
155   2014-05-25         87        80        72           67.0          66.0   
164   2014-06-03         90        83        75           72.0          70.0   
169   2014-06-08         90        83        76           72.0          70.0   
174   2014-06-13         89        79        68            NaN           NaN   
175   2014-06-14         87        92        73            NaN           NaN   
176   2014-06-15         91        83        74            NaN           NaN   
177   2014-06-16         92        84        75            NaN           NaN   
182   2014-06-21         92        84        75           74.0          72.0   
183   2014-06-22         90        83        76           74.0          72.0   
188   2014-06-27         91        83        75           74.0          72.0   
189   2014-06-28         93        86        78           74.0          73.0   
248   2014-08-26        101        89        77           71.0          68.0   
278   2014-09-25         83        73        62           66.0          62.0   
280   2014-09-27         80        74        68           69.0          66.0   
...          ...        ...       ...       ...            ...           ...   
1101  2016-12-26         84        77        70           70.0          69.0   
1102  2016-12-27         76        69        62           68.0          58.0   
1105  2016-12-30         55        52        49           34.0          23.0   
1106  2016-12-31         77        64        51           61.0          51.0   
1107  2017-01-01         73        63        53           60.0          52.0   
1125  2017-01-19         71        62        52           55.0          52.0   
1134  2017-01-28         61        50        39           37.0          31.0   
1140  2017-02-03         53        49        45           37.0          34.0   
1142  2017-02-05         72        62        51           64.0          59.0   
1143  2017-02-06         78        73        67           67.0          65.0   
1150  2017-02-13         78        71        63           64.0          55.0   
1163  2017-02-26         72        57        42           64.0          50.0   
1164  2017-02-27         82        74        66           66.0          65.0   
1171  2017-03-06         84        76        68           67.0          66.0   
1172  2017-03-07         77        69        61           67.0          51.0   
1173  2017-03-08         76        68        60           64.0          50.0   
1174  2017-03-09         80        74        67           66.0          64.0   
1175  2017-03-10         83        73        62           65.0          62.0   
1209  2017-0

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [11]:
# Your code here
print(austin.shape[0],missing_values.shape[0])

1319 136


#### Calculate the ratio of missing rows to total rows

In [12]:
# Your code here
ratio_missing_rows=(100*missing_values.shape[0])/austin.shape[0]
ratio_missing_rows

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [13]:
# Your code here
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [14]:
# Your code here
ratio_missing_precipit=(austin.isna().sum().PrecipitationSumInches*100)/austin.shape[0]
ratio_missing_precipit

9.401061410159212

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [15]:
# Your code here 
austin.drop('PrecipitationSumInches', inplace=True,axis=1)

# Print `austin` to confirm the column is indeed removed
austin

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
5     2013-12-26         57        48        39           39.0          36.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
8     2013-12-29         64        50        36           49.0          41.0   
9     2013-12-30         44        40        35           31.0          26.0   
10    2013-12-31         55        46        36           31.0          28.0   
11    2014-01-01         69        54        39           51.0          42.0   
12    2014-01-02         55        44        33           39.0          26.0   
13    2014-01-03         58        43        28           37.0          22.0   
14    2014-01-04         71        57        42           55.0          48.0   
15    2014-01-05         59        47        34           54.0          32.0   
16    2014-01-06         36        29        22           15.0           8.0   
17    2014-01-07         48        35        22           29.0          11.0   
18    2014-01-08         53        47        40           51.0          45.0   
19    2014-01-09         70        62        53           60.0          55.0   
20    2014-01-10         72        65        57           64.0          61.0   
21    2014-01-11         75        62        48           54.0          37.0   
22    2014-01-12         67        57        46           58.0          47.0   
23    2014-01-13         71        57        43           60.0          36.0   
24    2014-01-14         76        57        37           31.0          25.0   
25    2014-01-15         64        52        40           29.0          23.0   
26    2014-01-16         72        56        40           31.0          27.0   
27    2014-01-17         64        54        43           31.0          20.0   
28    2014-01-18         70        54        37           41.0          33.0   
29    2014-01-19         72        59        45           41.0          30.0   
...          ...        ...       ...       ...            ...           ...   
1289  2017-07-02        100        89        78           75.0          72.0   
1290  2017-07-03         99        88        76           74.0          71.0   
1291  2017-07-04        101        90        78           74.0          72.0   
1292  2017-07-05         99        89        78           74.0          70.0   
1293  2017-07-06        100        89        77           74.0          71.0   
1294  2017-07-07         99        87        75           74.0          71.0   
1295  2017-07-08        100        88        75           71.0          67.0   
1296  2017-07-09        101        89        77           71.0          67.0   
1297  2017-07-10        100        88        76           72.0          65.0   
1298  2017-07-11        101        89        77           73.0          69.0   
1299  2017-07-12        101        89        77           73.0          70.0   
1300  2017-07-13        102        90        77           74.0          70.0   
1301  2017-07-14        101        89        77           74.0          69.0   
1302  2017-07-15        103        87        71           73.0          70.0   
1303  2017-07-16         99        84        68           71.0          69.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1305  2017-07-18        100        88        76           74.0          70.0   
1306  2017-07-19        101        89        77           74.0          70.0   
1307  2017-0

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [16]:
# Your code here
austin_fixed=austin.interpolate(inplace=False)
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                Events  
0  Rain , Thunderstorm  
1                       
2                       
3                       
4

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [17]:
# Your code here
display(austin_fixed.isna().sum())
display(austin.isna().sum())

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [18]:
# Your code here:
austin['Events'].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Fog , Thunderstorm             1
Rain , Snow                    1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [19]:
# Your answer:
'''the max number of events is 3'''

'the max number of events is 3'

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [20]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for i in event_list:
    austin_fixed[i]=0
# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
5     2013-12-26         57        48        39           39.0          36.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
8     2013-12-29         64        50        36           49.0          41.0   
9     2013-12-30         44        40        35           31.0          26.0   
10    2013-12-31         55        46        36           31.0          28.0   
11    2014-01-01         69        54        39           51.0          42.0   
12    2014-01-02         55        44        33           39.0          26.0   
13    2014-01-03         58        43        28           37.0          22.0   
14    2014-01-04         71        57        42           55.0          48.0   
15    2014-01-05         59        47        34           54.0          32.0   
16    2014-01-06         36        29        22           15.0           8.0   
17    2014-01-07         48        35        22           29.0          11.0   
18    2014-01-08         53        47        40           51.0          45.0   
19    2014-01-09         70        62        53           60.0          55.0   
20    2014-01-10         72        65        57           64.0          61.0   
21    2014-01-11         75        62        48           54.0          37.0   
22    2014-01-12         67        57        46           58.0          47.0   
23    2014-01-13         71        57        43           60.0          36.0   
24    2014-01-14         76        57        37           31.0          25.0   
25    2014-01-15         64        52        40           29.0          23.0   
26    2014-01-16         72        56        40           31.0          27.0   
27    2014-01-17         64        54        43           31.0          20.0   
28    2014-01-18         70        54        37           41.0          33.0   
29    2014-01-19         72        59        45           41.0          30.0   
...          ...        ...       ...       ...            ...           ...   
1289  2017-07-02        100        89        78           75.0          72.0   
1290  2017-07-03         99        88        76           74.0          71.0   
1291  2017-07-04        101        90        78           74.0          72.0   
1292  2017-07-05         99        89        78           74.0          70.0   
1293  2017-07-06        100        89        77           74.0          71.0   
1294  2017-07-07         99        87        75           74.0          71.0   
1295  2017-07-08        100        88        75           71.0          67.0   
1296  2017-07-09        101        89        77           71.0          67.0   
1297  2017-07-10        100        88        76           72.0          65.0   
1298  2017-07-11        101        89        77           73.0          69.0   
1299  2017-07-12        101        89        77           73.0          70.0   
1300  2017-07-13        102        90        77           74.0          70.0   
1301  2017-07-14        101        89        77           74.0          69.0   
1302  2017-07-15        103        87        71           73.0          70.0   
1303  2017-07-16         99        84        68           71.0          69.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1305  2017-07-18        100        88        76           74.0          70.0   
1306  2017-07-19        101        89        77           74.0          70.0   
1307  2017-0

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [21]:
# Your code here
for i in event_list:
        austin_fixed[i]=austin_fixed['Events'].astype(str).str.contains(i,regex=True).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [22]:
# Your code here
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
5     2013-12-26         57        48        39           39.0          36.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
8     2013-12-29         64        50        36           49.0          41.0   
9     2013-12-30         44        40        35           31.0          26.0   
10    2013-12-31         55        46        36           31.0          28.0   
11    2014-01-01         69        54        39           51.0          42.0   
12    2014-01-02         55        44        33           39.0          26.0   
13    2014-01-03         58        43        28           37.0          22.0   
14    2014-01-04         71        57        42           55.0          48.0   
15    2014-01-05         59        47        34           54.0          32.0   
16    2014-01-06         36        29        22           15.0           8.0   
17    2014-01-07         48        35        22           29.0          11.0   
18    2014-01-08         53        47        40           51.0          45.0   
19    2014-01-09         70        62        53           60.0          55.0   
20    2014-01-10         72        65        57           64.0          61.0   
21    2014-01-11         75        62        48           54.0          37.0   
22    2014-01-12         67        57        46           58.0          47.0   
23    2014-01-13         71        57        43           60.0          36.0   
24    2014-01-14         76        57        37           31.0          25.0   
25    2014-01-15         64        52        40           29.0          23.0   
26    2014-01-16         72        56        40           31.0          27.0   
27    2014-01-17         64        54        43           31.0          20.0   
28    2014-01-18         70        54        37           41.0          33.0   
29    2014-01-19         72        59        45           41.0          30.0   
...          ...        ...       ...       ...            ...           ...   
1289  2017-07-02        100        89        78           75.0          72.0   
1290  2017-07-03         99        88        76           74.0          71.0   
1291  2017-07-04        101        90        78           74.0          72.0   
1292  2017-07-05         99        89        78           74.0          70.0   
1293  2017-07-06        100        89        77           74.0          71.0   
1294  2017-07-07         99        87        75           74.0          71.0   
1295  2017-07-08        100        88        75           71.0          67.0   
1296  2017-07-09        101        89        77           71.0          67.0   
1297  2017-07-10        100        88        76           72.0          65.0   
1298  2017-07-11        101        89        77           73.0          69.0   
1299  2017-07-12        101        89        77           73.0          70.0   
1300  2017-07-13        102        90        77           74.0          70.0   
1301  2017-07-14        101        89        77           74.0          69.0   
1302  2017-07-15        103        87        71           73.0          70.0   
1303  2017-07-16         99        84        68           71.0          69.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1305  2017-07-18        100        88        76           74.0          70.0   
1306  2017-07-19        101        89        77           74.0          70.0   
1307  2017-0

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [23]:
# Your code here
austin_fixed=austin_fixed.drop('Events', axis=1)
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  \
0                 2.0         20.0         4.0         31.0     0    0     1   
1                 5.0         16.0         6.0         25.0     0    0     0   
2                10.0          8.0         3.0         12.0     0    0     0   
3                 7.0         12.0         4.0         20.0     0    0     0   
4                 7.0         10.0         2.0         16.0     0    0     0   

   Thunderstorm  
0             1  
1             0  
2             0  
3             0  
4             0

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [24]:
# Your code here
from datetime import datetime
austin_fixed["Date"] = [datetime.strptime(str(i), '%Y-%m-%d').date() for i in austin_fixed["Date"]]
austin_fixed["Date"] = [i.toordinal() for i in austin_fixed["Date"]]

#### Print `austin_fixed` to check your `Date` column.

In [25]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  \
0                 2.0         20.0         4.0         31.0     0    0     1   
1                 5.0         16.0         6.0         25.0     0    0     0   
2                10.0          8.0         3.0         12.0     0    0     0   
3                 7.0         12.0         4.0         20.0     0    0     0   
4                 7.0         10.0         2.0         16.0     0    0     0   

   Thunderstorm  
0             1  
1             0  
2             0  
3             0  
4             0

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [26]:
# Your code here:
cols=['TempHighF','TempLowF','DewPointHighF','DewPointAvgF','DewPointLowF','HumidityHighPercent','HumidityAvgPercent','HumidityLowPercent','SeaLevelPressureHighInches','SeaLevelPressureAvgInches','SeaLevelPressureLowInches','VisibilityHighMiles','VisibilityAvgMiles','VisibilityLowMiles','WindHighMPH','WindAvgMPH','WindGustMPH','Snow','Fog','Rain','Thunderstorm']
x=austin_fixed[cols]
y=austin_fixed['TempAvgF']

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [27]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [28]:
#Your code here:
train_test_split(x,y,test_size=0.2)

[      TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
 1232         91        54           53.0          49.0          46.0   
 1286         99        75           75.0          73.0          70.0   
 475          74        56           68.0          53.0          47.0   
 491          86        69           72.0          69.0          63.0   
 700          66        41           63.0          47.0          25.0   
 586         101        75           71.0          66.0          60.0   
 33           56        27           42.0          29.0          20.0   
 245         101        75           73.0          69.0          62.0   
 899          92        69           70.0          65.0          63.0   
 215          94        74           77.0          72.0          68.0   
 480          83        54           65.0          56.0          50.0   
 1065         67        35           36.0          32.0          27.0   
 439          47        31           32.0          

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [29]:
# Your code here:
ts_rows=round(((int(austin_fixed.shape[0]))*80)/100)
ts_rows

1055

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [38]:
# Your code here:
X_ts_train=x[:ts_rows+1]
X_ts_test=x[ts_rows+1:]

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [39]:
# Your code here:
Y_ts_train=y[:ts_rows+1]
Y_ts_test=y[ts_rows+1:]

# Bonus Challenge 2

As explained in the README, the main purpose of this lab is to show you the typical process of preparing data for machine learning which sometimes takes up 90% of your time. Data cleaning is a valuable skill to learn and you need to be proficient at various techniques including the ones we showed you above as well as others you'll learn in the future. In the real world this skill will help you perform your job successfully and efficiently.

Now that we're done with data praparation, if you want to expeirence what you'll do in the rest 10% of your time, let's make the final leap.

We will use scikit-learn's [*Support Vector Machines*](https://scikit-learn.org/stable/modules/svm.html) to compute the fit of our training data set, the test on our test data set.

#### In the cell below, import `svm` from `sklearn`:

In [32]:
# Your code here


#### Now, call `svm.SVC.fit()` on `X_train` and `y_train`. Assign the returned value to a variable called `clf` which stands for *classifier*. Then obtain the test score for `X_test` and `y_test` by calling `clf.score()`.

In [33]:
# Your code here


#### You now see the model fit score of your test data set. If it's extremely low, it means your selected model is not a good fit and you should try other models.

#### In addition to fitting `X_train`, `y_train`, `X_test`, and `y_test`, you can also fit `X_ts_train`, `y_ts_train`, `X_ts_test`, and `y_ts_test` if you completed Bonus Challenge 1.

In [34]:
# Your code here


#### We hope you have learned a lot of useful stuff in this lab!